pkg/objstore/objstore.go

```diff
@@ -137,12 +140,38 @@ func TryToGetSize(r io.Reader) (int64, error) {
 		return fileInfo.Size(), nil
 	case *bytes.Buffer:
 		return int64(f.Len()), nil
+	case *bytes.Reader:
+		// Returns length of unread data only.
+		return int64(f.Len()), nil
 	case *strings.Reader:
 		return f.Size(), nil
+	case ReaderWithSize:
+		return f.Size()
 	}
 	return 0, errors.Errorf("unsupported type of io.Reader: %T", r)
 }
 
+// ReaderWithSize is a Reader that can also return size of the read data.
+type ReaderWithSize interface {
```

Comment:
I think @pstibrany this makes sense if we would replace the `io.Reader` with this. Otherwise, I wonder if it would be not easier to just have `Sizer` interface and cast to that in `TryToGetSize` (: WDYT?

Reply:
I would be happy I think to accept `ReaderWithSize` explicitly by our Upload - It will being some  complexity to the user, but at least avoid efficiency issues.

Reply:
> Otherwise, I wonder if it would be not easier to just have `Sizer` interface and cast to that in `TryToGetSize` (: WDYT?

I considered this approach, but it felt very generic. I wanted to express that this is really a Reader which is aware of size of (possibly unread) data. If it was defined only as Sizer, it would be weird to refer to readers’ unread data in the comment. If you however still prefer that approach, I am fine with that.

> I would be happy I think to accept `ReaderWithSize` explicitly by our Upload - It will being some complexity to the user, but at least avoid efficiency issues.

I’m not in favor of this for few reasons:
- [amount of code using Thanos objstore library](https://pkg.go.dev/github.com/thanos-io/thanos/pkg/objstore?tab=importedby), that would need to change. There is also some private code not shown on that page. I’m not sure the benefits outweigh causing so much trouble :)
- Interface is defined such that `Size` can return error. What would Upload do in such case? Fail? Or would it be implementation specific? If it doesn’t fail, why require it in the first place?
- Typical object storage doesn’t need this information in advance, so requiring it by API seems counter-intuitive to me.

Reply:
I forgot to thank you for your review! :)

Reply:
This is an amazing discussion to have. Let's go through this, will try to be peeking on this discussion more often (feel free to ping me on slack too). 

> I wanted to express that this is really a Reader who is aware of the size of (possibly unread) data.

Does it matter for purpose of this interface and the caller needs to understand if it's a reader/writer? I think this is actually a greater pattern and consistent with other Go code like e.g `http.Flusher` -> it's not `ReaderFlusher`.  I would vote for Sizer yes.

> Interface is defined such that Size can return error. What would Upload do in such case? Fail? Or would it be implementation specific? If it doesn’t fail, why require it in the first place?

We have exactly the same question now with Sizer or ReaderSizer casting, no?

> amount of code using Thanos objstore library, that would need to change. There is also some private code not shown on that page. I’m not sure the benefits outweigh causing so much trouble :)
> Typical object storage doesn’t need this information in advance, so requiring it by API seems counter-intuitive to me.

Hm, but actually we really really care to provide it otherwise we are doing thins inefficiently. And that amount of private or public Go has to be aware of this and leverage type mechanism instead of run time best effort casting. But you are right, that's a bigger change, happy with sizer for now 🤗

Reply:
Good question. Let's iterate slowly (: First optional, but error should fail Upload IMO